orc8r/cloud/go/tools/migrations/m010_default_apns/main.go

```diff
@@ -0,0 +1,452 @@
+	}
+
+	assocs := map[ent][]ent{}
+	for rows.Next() {
+		a := ent{typ: subscriberEntType}
+		pkB, typeB := &sql.NullString{}, &sql.NullString{}
+		err = rows.Scan(&a.graphID, &a.pk, pkB, typeB)
+		if err != nil {
+			return nil, errors.Wrap(err, "getSubscribersMissingAPN: scan subscriber")
+		}
+		assocs[a] = append(assocs[a], ent{pk: pkB.String, typ: typeB.String})
+	}
+	err = rows.Err()
+	if err != nil {
+		return nil, errors.Wrap(err, "getSubscribersMissingAPN: SQL rows error")
+	}
+
+	var subsMissingAPN []ent
+	for sub, ents := range assocs {
+		hasAPN := false
+		for _, ent := range ents {
+			if ent.typ == apnEntType {
+				hasAPN = true
+			}
+		}
+		if !hasAPN {
+			subsMissingAPN = append(subsMissingAPN, sub)
+		}
+	}
+
+	return subsMissingAPN, nil
+}
+
+func mapSubscribersToAPN(tx *sql.Tx, builder sqorc.StatementBuilder, apnPK string, subscriberPKs []string) error {
+	b := builder.Insert(entityAssocTable).Columns(aFrCol, aToCol)
+	for _, subPK := range subscriberPKs {
+		b = b.Values(subPK, apnPK)
+	}
+	sqlStr, args, _ := b.ToSql()
+	glog.Infof("[RUN] %s %v", sqlStr, args)
+	_, err := b.RunWith(tx).Exec()
+	if err != nil {
+		return errors.Wrap(err, "mapSubscribersToAPN: insert error")
+	}
+
+	return nil
+}
+
+// mergeGraphs merges the graphs for all passed graph IDs.
+// Assumes gids is non-empty.
+func mergeGraphs(tx *sql.Tx, builder sqorc.StatementBuilder, network string, gids []string) error {
+	gids = funk.UniqString(gids)
+	sort.Strings(gids)
+	mergedGID, oldGIDs := gids[0], gids[1:]
+
+	isOldGID := squirrel.Or{}
+	for _, gid := range oldGIDs {
+		isOldGID = append(isOldGID, squirrel.Eq{entGidCol: gid})
+	}
+
+	b := builder.
+		Update(entityTable).
+		Set(entGidCol, mergedGID).
+		Where(
+			squirrel.And{
+				squirrel.Eq{entNidCol: network},
+				isOldGID,
+			},
+		)
+	sqlStr, args, _ := b.ToSql()
+	glog.Infof("[RUN] %s %v", sqlStr, args)
+	_, err := b.RunWith(tx).Exec()
+	if err != nil {
+		return errors.Wrap(err, "mergeGraphs: update error")
+	}
+
+	return nil
+}
+
+// verifyMigration loads all subscribers into memory, and checks that they
+// each have an APN assoc.
+func verifyMigration() error {
+	registry.MustPopulateServices()
+
+	nids, err := configurator.ListNetworkIDs()
+	if err != nil {
+		return err
+	}
+
+	for _, nid := range nids {
+		subs, err := configurator.LoadAllEntitiesInNetwork(nid, subscriberEntType, configurator.EntityLoadCriteria{LoadAssocsFromThis: true})
+		if err != nil {
+			return err
+		}
+
+		i := 0
+		for _, sub := range subs {
+			apns := funk.Chain(sub.Associations).
+				Filter(func(tk storage.TypeAndKey) bool { return tk.Type == apnEntType }).
+				Map(func(tk storage.TypeAndKey) string { return tk.Key }).
+				Value().([]string)
+
+			if len(apns) == 0 {
+				return fmt.Errorf("subscriber %s has no APN assocs", sub.Key)
+			}
+
+			if i < numManualVerificationsToLog {
+				glog.Infof("Subscriber %v has APN assocs %v", sub, apns)
+			}
+			i += 1
+		}
+	}
+
+	return nil
+}
```

Comment:
I think it makes sense to sanity check the graphs for the default apns are correct. Load the default APN entity, then load its graph, and just verify that we get back exactly the set of subscribers that the APN has assocs to in the graph result